### PR TITLE
Lin Out_channel shrink cleanup

### PR DESCRIPTION
**Pull request by jmid**
This PR contains a range of adjustments to address the long shrinking sequences reported in #378:
- bf617290641d234f951600d33fa455d42256f1da and d5ff92a21ccbce6e84eb75d7d58e39cf10a8290d disable `char` element shrinking inside `bytes` and `string`s for the test
- 2433827343b1a80030dae97943fe969403cd70fe removes the `lift` combinator to reduce allocations
- ba20a6f8aee7b7aaf2a498d00345c4e2cbce2f00 adjusts the `cmd` frequencies
- 6c8db20adee709998417dd73da2ece6ff792512f drops `flush_all` entirely
- Finally e1a1de31a6cff43649e27810d804fafce3e6f256 reduces overall `Lin` allocations during `check_seq_cons` and should thus benefit other `Lin` tests too.

Overall, this makes the test behaviour reasonable. Below I include a CI summary table from running 20 focused tests across the 2 CI systems (I note runs taking more than 1000s as "outliers"). Observations:
- There's a trend of more outliers+timeouts after 5.0 - extra GC activity on +5.1.0 has been confirmed with `perf` and `olly gc-stats` measurements
- Disabling the test on macOS should be kept for now, an explanation would be nice though
- There was a crazy linux 5.1 debug 12514.3s outlier!

Overall the PR offers improvements, not the final word. They should help reduce false CI alarms though.

The last two commits 460b97c88439eac44a23eb81856977b19066d4cd and f7223d66b17148e91c90887a7d8d56fec318fe9b update the plain `Lin` `Out_channel` test to move it closer feature-wise to the `Lin DSL` version. These are steps towards creating an standalone reproducer for the extra +5.1.0 GC activity.

| | Target            | Outliers                                        | Timeout | Remarks |
|-|-------------------|-------------------------------------------------|---------|---------|
|✅|Linux 5.1          | -
|✅|Linux trunk        | 1973.0s
|❌|Linux 5.1 debug    | 12514.3s                                        | 12th run
|✅|Linux trunk debug  | 1462.0s 2350.7s
|✅|32bit trunk        | -
|✅|32bit 5.1          | -
|✅|Bytecode 5.1       | -
|✅|Bytecode trunk     | -
|  | | |
|✅|linux-arm64-5.0    | - |
|✅|linux-arm64-5.1    | 2367.5s 1150.1s 1360.5s 3558.0s
|✅|linux-arm64-5.2    | 1426.0s 1416.6s 1435.6s 1160.6s
| | | |
|✅|linux-s390x-5.1    | 1493.5s
|✅|linux-s390x-5.2    | -
| | | |
|❌|linux-ppc64le-5.2  | 2497.0s 1434.2s 3345.5s 3099.9s 1852.6s 2503.3s | 11th run
| | | |
|❌|macOS 5.1          | 7728.8s 1371.7s                                 | 3rd run
|❌|macOS trunk        | -                                               | 2nd run
|✅|macos-arm64-5.0    | 2090.3s
|❌|macOS-arm64-5.1    | 3017.1s 3278.3s 2001.6s 1961.0s 1866.6s 2803.1s | 10th run | 6/9 w/no counterexample!
|❌|macOS-arm64-5.2    | 3988.2s 3447.5s 3344.8s 2651.1s 3579.0s         | 6th run  | 5/5 w/no counterexample!
| | | |
|✅|Windows 5.1        | 1030.1s 2523.1s
|✅|Windows trunk      | 2471.4s 1139.8s 2522.6s
|✅|Win bytecode 5.1   | 1391.8s 3730.3s
|✅|Win bytecode trunk | -
| | | |
|✅|Cygwin 5.1         | 1004.0s 1206.1s
|✅|Cygwin trunk       | 1010.6s

### CODE REVIEW

**Comment by jmid** [comment]
Oh wow :open_mouth: linux-arm64 5.2 had a 16443.2s `Lin DSL Out_channel test with Domain` run, yet somehow still managed to complete within the timeout limit!

File: lib/lin.ml [hunk 1]
```diff
@@ -113,7 +116,7 @@ struct
           ||
           (* rerun to get seq_sut to same cmd branching point *)
           (let seq_sut' = Spec.init () in
-           let _ = interp_plain seq_sut' (List.rev seq_trace) in
+           let _ = interp_plain_ignore seq_sut' (List.rev seq_trace) in
```
Comment (shym) [inline lib/lin.ml:119]: Maybe `interp_plain_ignore seq_sut' (List.rev seq_trace);` rather than `let _ = ... in`?

**Review by shym — APPROVED** [review]
Looks good to me! :-)

File: lib/lin.ml (hunk 1 above)
Comment (jmid) [reply lib/lin.ml]: Fair point - addressed in 1b50d2f

**Comment by jmid** [comment]
CI summary for f7223d6 from yesterday:
- 3 Windows threadomain failure #203
  - bytecode 5.1 - new lock failure variant (already noted in #203)
  - bytecode trunk/5.2 crash and trunk/5.2 triggered regular crashes
- 3 Cygwin part2 timeouts #384 #338
  - 5.1 - during a 5961.1s `Lin CList int64 test with Thread`
  - trunk (PR) during a 8343.3s `Lin ref int64 test with Thread`
  - trunk (push) during a 4730.3s `Lin CList int64 test with Thread`
- linux-ppc64le-5.2 failed with `SEGV` during sequential `STM Array` and `Bytes` tests #380

For the Cygwin failures, I get suspicious that the slowness is affecting the ability to trigger test failures and thereby cause needlessly long shrink runs :thinking:

**Comment by jmid** [comment]
CI summary for bb94d34:

- multicoretest-ci runs were cancelled
- Cygwin trunk part 1 timeout during `Dynlink` test #307 
- MingW bytecode 5.1 and MingW trunk timeouts during `threadomain` #203 

So 3/23 failures - but none of them being false positives! :smiley:

**Comment by jmid** [comment]
CI summary for b9509bb
- MingW bytecode trunk timeout during threadomain #203
- Cygwin 5.1 part 2 timeout during a 3016.7s `Lin CList int64 test with Thread`, incl. a 5144.7s `Lin ref int64 test with Thread` #384 #338
- linux-ppc64le-5.2 SEGV'ed during sequential `STM Bytes` and `Float Array` tests #380
- macos-arm64-5.1 failed with an opam switch 5.1.0~rc2 build failure

So, 4/52 failures with only 1 of them being noise (Cygwin slowness is a genuine issue) and 1 being container related.

I'll merge :smiley:

**Pull request merged**